### DISCUSSION TITLE
tableofcontents directive in document with no descendants [etoc.tableofcontents]

### DISCUSSION THREAD

**Discussion by giswqs** [executablebooks/meta#704] [Q&A] [answered]
I am trying to insert a table of contents into a page built from myst using the `{tableofcontents}` directive, but got a warning saying that `tableofcontents directive in document with no descendants [etoc.tableofcontents]`. The table of contents does not show up on the HTML page. Any advice?

![image](https://user-images.githubusercontent.com/5016453/160620053-ea271879-b2c0-437b-ac30-c48edcd8b382.png)

**Comment by choldgraf** [comment, marked as answer]
The tableofcontents directive is for listing sub-pages underneath the current page. Is this true in your case? Eg are there files listed under "sections" in your toc.yml file?

**Reply by giswqs** [reply]
I see. I was trying to add a table of contents of the current page to the beginning of the page. The headings are not listed in the toc.yml file. Thanks for clarifying.

**Reply by choldgraf** [reply]
For that I believe you can just use {contents} - does that work?

**Reply by giswqs** [reply]
Great to know this. It indeed works. The only issue is that it also includes the title (heading level 1 of the notebook). Can heading level 1 be removed from the TOC? 

![image](https://user-images.githubusercontent.com/5016453/160643311-6e3c71d2-71e1-46ba-a47a-d362856ebbd6.png)

**Reply by giswqs** [reply]
BTW, is it possible to control the max depth of TOC generated using the `{contents}` directive? For example, I only want the TOC to show heading levels up to 3. The max depth of the `{tableofcontents}` directive can be specified in the `_toc.yml`, but the `contents` directive does not seem to have this option.

**Reply by choldgraf** [reply]
Yep - two answers below:

**To remove the title** could you try adding the `:local:` argument? The `:local:` argument should only include the contents in the *current* section, and if you put this under the title, this should be all top-level sections.

**To limit the depth** could you try adding a `:depth:` argument? e.g. `:depth: 2`.

So something like:

````
# Page title

```{contents}
:local:
:depth: 2
```
````

**Reply by giswqs** [reply]
It works like charm. Thanks a lot for helping out. 

I think this would be useful to add to the documentation. It will help newcomers. 
https://jupyterbook.org/structure/toc.html

**Reply by choldgraf** [reply]
@giswqs I agree! Would you be willing to add a short PR? I am happy to review and edit as needed!

**Reply by giswqs** [reply]
Just submitted the PR. Thanks. https://github.com/executablebooks/jupyter-book/pull/1683
